File: Java/Hard/2573 Find the String with LCP.java
class Solution {
    public String findTheString(int[][] lcp) {
        int n = lcp.length;
        int[] s = new int[n];
        int charValue = 1;

        for (int i = 0; i < n; i++) {
            if (s[i] > 0) continue;
            if (charValue > 26) return "";
            for (int j = i; j < n; j++) {
                if (lcp[i][j] > 0) {
                    s[j] = charValue;
                }
            }
            charValue++;
        }

        for (int i = n - 1; i >= 0; i--) {
            for (int j = n - 1; j >= 0; j--) {
                int expected = (s[i] == s[j]) ? 
                               (i == n - 1 || j == n - 1 ? 1 : 1 + lcp[i + 1][j + 1]) : 0;
                if (lcp[i][j] != expected) return "";
            }
        }

        StringBuilder sb = new StringBuilder();
        for (int v : s) sb.append((char) ('a' + v - 1));
        return sb.toString();
    }
}

File: Python/Hard/2573 Find the String with LCP.py
class Solution:
    def findTheString(self, lcp: List[List[int]]) -> str:
        n = len(lcp)
        s = [0] * n
        char_val = 1
        
        for i in range(n):
            if s[i] > 0: continue
            if char_val > 26: return ""
            for j in range(i, n):
                if lcp[i][j] > 0:
                    s[j] = char_val
            char_val += 1
            
        for i in range(n - 1, -1, -1):
            for j in range(n - 1, -1, -1):
                expected = 0
                if s[i] == s[j]:
                    expected = 1 + (lcp[i+1][j+1] if i < n-1 and j < n-1 else 0)
                if lcp[i][j] != expected:
                    return ""
                    
        return "".join(chr(ord('a') + v - 1) for v in s)
